includes/class-wp-smime.php

```diff
@@ -229,4 +239,122 @@ public static function filterContentType ( $content_type ) {
+     * It's rather an attempt to make the best out of a very sub-optimal situation.
+     *
+     * @param array $files
+     */
+    private static function safeDelete( ...$files ) {
+
+        foreach ( $files as $f ) {
+
+            // clear stat cache to avoid falsely reported file status
+            clearstatcache();
+
+            if ( ! file_exists( $f ) || ! is_file( $f ) ) {
+                // nothing we can do about this
+                error_log( "File path is invalid!" );
+                continue;
+            }
+            if ( ! is_writable( $f ) ) {
+                error_log( "File is not writable!" );
+
+                // cant overwrite, try to delete it
+                if ( ! unlink( $f ) ) {
+                    error_log( "Could not delete file!" );
+                }
+                continue;
+            }
+
+            // try to delegate the work to coreutils 'shred' for better performance
+            // exists on almost all linux systems
+            // https://www.gnu.org/software/coreutils/manual/html_node/shred-invocation.html#shred-invocation
+            $handle = popen( "shred -fuz {$f}", "r" );
+
+            if ( -1 !== pclose( $handle ) ) {
+                clearstatcache();
+
+                // unfortunately we can't get the original exit code of 'shred'
+                if ( ! file_exists( $f ) ) {
+
+                    // assume shredding was successful,
+                    // continue with next file
+                    continue;
+                } else {
+                    error_log( "Shredding was not successful!" );
+                }
+            } else {
+                error_log( "Error while running 'shred'!" );
+            }
+
+            // determine file size
+            $size = filesize( $f );
+
+            // if FALSE, size must not be 0!
+            if ( false === $size || $size < 0 ) {
+                // set default size to 1 MiB
+                $size = 1 * 1024 * 1024;
+            }
+
+            // try to overwrite 3 times
+            for ( $i = 0; $i < 3; ++ $i ) {
+
+                // randomly increase size
+                try {
+                    if ( $size > 0 ) {
+                        $size += random_int( 1, $size * 2 );
+
+                    } else if ( 0 === $size ) {
+                        // file should be empty, no need to overwrite,
+                        // but random_bytes() expects non-zero value
+                        $size = 1;
+                    }
+                } catch ( Exception $e ) {
+                    error_log( "Could not generate secure integer!", $e );
+
+                    // fallback to insecure rand()
+                    $size += rand( 1, $size * 2 );
+                }
+
+                $bytes = array();
+
+                try {
+                    $bytes = random_bytes( $size );
+
+                } catch ( Exception $e ) {
+                    error_log( "Could not generate random bytes!", $e );
```

Comment:
As I understand it, it will never return insecure random bytes, but it can still fail and throw an exception if it can't generate a secure output for some unlikely reason - e.g. low entropy.

Should we return after catching and logging the exception or just ignore it?